### PR TITLE
feat: Add `set_code` method to `ProviderAPI`

### PR DESCRIPTION
**Pull request by sabotagebeats**
### What I did
added set-code in providers api

fixes: APE-146

### How I did it
as per the spec

Specification

In `ProviderAPI`, add method:

```
@raises_not_implemented
def set_code(address, AddressType, code: bytes):
    """
    Change the code of a smart contract, for development purposes.
    Test providers implement this method when they support it.
    """
```

Then, in ape-hardhat, we can support it be calling the RPC https://hardhat.org/hardhat-network/docs/reference#hardhat_setcode

### Checklist

- [x] All changes are completed
- [x] New test cases have been added
- [x] Documentation has been updated

### CODE REVIEW

File: src/ape/api/providers.py [hunk 1]
```diff
@@ -418,6 +418,18 @@ def __repr__(self) -> str:
...
+    @raises_not_implemented
+    def set_code(self, address: AddressType, code: bytes):
+        """
+        Change the code of a smart contract, for development purposes.
+        Test providers implement this method when they support it.
+
+        Args:
+            address (AddressType): An address on the network.
+            code(bytes): The new bytecode
+
```
Comment (unparalleled-js) [inline src/ape/api/providers.py:430]:
```suggestion
```

File: src/ape/api/providers.py [hunk 2]
```diff
@@ -418,6 +418,18 @@ def __repr__(self) -> str:
...
 
+    @raises_not_implemented
+    def set_code(self, address: AddressType, code: bytes):
+        """
+        Change the code of a smart contract, for development purposes.
+        Test providers implement this method when they support it.
+
+        Args:
+            address (AddressType): An address on the network.
+            code(bytes): The new bytecode
```
Comment (unparalleled-js) [inline src/ape/api/providers.py:429]:
```suggestion
            code(bytes): The new bytecode.
```

**Review by unparalleled-js — COMMENTED** [review]
Awesome.

Before merging, let's add support to `ape-hardhat` as well as tests there, to prove the API method suffices in its signature.

**Comment by sabotagebeats** [comment]
> Awesome.
> 
> Before merging, let's add support to `ape-hardhat` as well as tests there, to prove the API method suffices in its signature.

what do we need to do to add `ape-hardhat` support for this feature?

File: src/ape/types/__init__.py [hunk 3]
```diff
@@ -31,6 +31,11 @@
 as well as a block number or hash (HexBytes).
 """
 
+ContractCode = Union[str, bytes, HexBytes]
+"""
+An ID that represents contract code, which can be represented in string, bytes, or HexBytes.
```
Comment (unparalleled-js) [inline src/ape/types/__init__.py:36]: It is not really an ID here

```suggestion
A type that represents contract code, which can be represented in string, bytes, or HexBytes.
```

**Review by unparalleled-js — DISMISSED** [review]

File: src/ape/api/providers.py [hunk 4]
```diff
@@ -180,7 +180,7 @@ def get_balance(self, address: str) -> int:
         """
 
     @abstractmethod
-    def get_code(self, address: str) -> bytes:
+    def get_code(self, address: AddressType) -> bytes:
```
Comment (unparalleled-js) [inline src/ape/api/providers.py:183]:
```suggestion
    def get_code(self, address: AddressType) -> ContractCode:
```

File: src/ape/api/providers.py [hunk 5]
```diff
@@ -180,7 +180,7 @@ def get_balance(self, address: str) -> int:
...
         """
         Get the bytes a contract.
```
Comment (unparalleled-js) [inline src/ape/api/providers.py:186]: The arg doc needs to be updated:

```python
Args:
    address (``AddressType``): The address of the contract.

Returns:
    :class:`~ape.types.ContractCode`: The contract bytecode.
```

File: src/ape/api/providers.py (hunk 4 above)
Comment (sabotagebeats) [reply src/ape/api/providers.py]: I think this should be resolved in newest commit.

File: src/ape/api/providers.py (hunk 5 above)
Comment (sabotagebeats) [reply src/ape/api/providers.py:186]: I think this should be resolved in newest commit.

File: src/ape/api/providers.py [hunk 6]
```diff
@@ -428,6 +428,17 @@ def __repr__(self) -> str:
...
 
+    @raises_not_implemented
+    def set_code(self, address: AddressType, code: ContractCode) -> bool:
+        """
+        Change the code of a smart contract, for development purposes.
+        Test providers implement this method when they support it.
+
+        Args:
+            address (AddressType): An address on the network.
+            code(bytes): The new bytecode.
```
Comment (unparalleled-js) [inline src/ape/api/providers.py:439]:
```suggestion
            code (:class:`~ape.types.ContractCode`): The new bytecode.
```

**Review by unparalleled-js — COMMENTED** [review]
1 issue in doc str

**Review by unparalleled-js — APPROVED** [review]

**Comment by sabotagebeats** [comment]
@unparalleled-js is there any more changes that need to be made or can this be merged?

**Pull request merged**